libnet/server.c

```diff
@@ -0,0 +1,302 @@
+				if(i == master_socket_fd) {
+					int client_fd;
+					if((client_fd = handle_client_init(master_socket_fd)) > 0) {
+						add_fd_to_select(client_fd);
+					}
+				} else if(i == selfpipe_fd) {
+					char ch;
+					log_info1("Got interrupted by selfpipe");
+					check1(read(selfpipe_fd, &ch, 1) >= 0, "selfpipe read");
+				} else {
+					handle_client_input(i);
+				}
+			}
+		}
+
+	}
+	close_socket(master_socket_fd);
+
+	return 0;
+error:
+	return -1;
+}
+
+static void* libnet_main_pthread_wrapper(void * args) {
+	//TODO(florek) return value!
+	libnet_main((char *)args);
+	return 0;
+}
+
+pthread_t libnet_thread;
+
+/*
+static int setup_signal_handler() {
+	struct sigaction sa;
+	sa.sa_handler = SIG_IGN;
+	sa.sa_flags = 0;
+
+	return sigaction(SIGPIPE, &sa, 0);
+}
+*/
+
+bool libnet_thread_start(char *address) {
+	//check1(setup_signal_handler() == 0, "Signal handler setup");
+	//TODO(florek) better error handling?
+	check1(!pthread_create(&libnet_thread, 0, libnet_main_pthread_wrapper, &address),
+			"pthread_create libnet_thread");
+	return true;
+error:
+	return false;
+}
+
+static bool libnet_thread_shutdown() {
+	exiting = true;
+	check1(notify_selfpipe(selfpipe_write_end) >= 0, "selfpipe write");
+	//TODO(florek) error handling
+	//TODO(florek) return val hadnling
+	check1(!pthread_join(libnet_thread, 0), "pthread_join libnet_thread");
+	return true;
+error:
+	return false;
+}
+
+#ifdef _DEBUG
+void libnet_debug_dump_client_info() {
+	for(unsigned i = 0; i < MAX_CLIENT_NUMBER; i++) {
+		client_info *c = clients[i];
+		debug("%d %s %s %d", c->fd, state_to_string(c->state), read_status_to_string(c->read_status), c->retries);
+	}
+}
+#endif
+
+bool libnet_send(unsigned char tag, size_t length, unsigned char *value) {
+	bool success = true;
+
+	for(unsigned i = 0; i < MAX_CLIENT_NUMBER; i++) {
+		client_info *c = clients + i;
+
+		success &= send_tag(c, tag, length, value);
+	}
+
+	return success;
+}
+
+ssize_t libnet_wait_for_tag(unsigned char tag, char *buffer, size_t length) {
+	int error = ENOTAG;
+	check1(wait_for_tag(tag), "libnet_wait_for_tag wait_for_tag");
+
+	tlv** message_queue = get_message_queue();
+
+	unsigned tag_message;
+	for(tag_message = 0; tag_message < MAX_MESSAGE_QUEUE &&
+		message_queue[tag_message] && message_queue[tag_message]->tag != tag;
+		tag_message++);
+
+	error = EQUEUE;
+	check1(message_queue[tag_message]->tag == tag, "Message queue inconsistent");
+
+	tlv* msg = message_queue[tag_message];
+	message_queue[tag_message] = 0;
+
+	error = ESIZE;
+	check1(msg->length <= length, "Buffer too small");
+
+	memcpy(buffer, msg->value, msg->length);
+
+	ssize_t message_length;
+	if(msg->length < SSIZE_MAX) {
+		message_length = (ssize_t) msg->length;
+	} else {
+		log_err1("Message longer than SSIZE_MAX");
+	}
+	free(msg);
+
+	return message_length;
+error:
+	return -error;
+}
+
+int main(int argc, char *argv[]) {
+
+	log_info1("Starting server thread");
+	libnet_thread_start(0);
```

Comment:
main function won't be used when building as a library and is here just for convinience 
( so I can run `gcc libnet_client.a -o client` and get client executable that just connects to predefined server, same for server). 
This code won't be in release executable.

It's the client/server that will be calling `libnet_thread_start` et al and it might want to know whether initialization was successful.